src/platform/qt/test/suite.h

```diff
@@ -0,0 +1,23 @@
+/* Copyright (c) 2013-2022 Jeffrey Pfau
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+#pragma once
+
+#define QT_NO_OPENGL
+
+#include <QtTest>
+#include <QList>
+
+struct TestRegistrar {
+	virtual int main(int argc, char** argv) = 0;
+};
+
+extern QList<TestRegistrar*> registeredTests;
+
+#define M_REGISTER_QT_TEST(Class) \
+	static struct Class ## Registrar : TestRegistrar { \
+		QTEST_MAIN(Class) \
+		Class ## Registrar() { registeredTests << this; } \
```

Comment:
There appears to be an initialization order conflict on my end whereby this function gets called before `registeredTests`'s constructor gets called, I think. It's really not clear what's going on, but it's crashing with a null `this`:

```
#4  0x000055555557fb35 in QList<TestRegistrar*>::append (this=0x555555599510 <registeredTests>, t=@0x7fffffffd8d0: 0x5555555994f8 <registrar_LibraryModelTest>) at /usr/include/qt/QtCore/qlist.h:622
622         if (d->ref.isShared()) {
(gdb) p this->d
$8 = (QListData::Data *) 0x0
```

Reply:
My suggestion is convert it into something that doesn't depend on constructors: a POD linked list.

Here's an example impl that works on my end and lets me run the tests:

```patch
diff --git a/src/platform/qt/test/main.cpp b/src/platform/qt/test/main.cpp
index 28a422415..2bf59edad 100644
--- a/src/platform/qt/test/main.cpp
+++ b/src/platform/qt/test/main.cpp
@@ -5,11 +5,11 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 #include "suite.h"

-QList<TestRegistrar*> registeredTests;
+TestRegistrar* TestRegistrar::first = nullptr;

 int main(int argc, char** argv) {
        int err = 0;
-       for (TestRegistrar* test : registeredTests) {
+       for (TestRegistrar* test = TestRegistrar::first; test; test = test->next) {
                int result = test->main(argc, argv);
                if (result && !err) {
                        err = result;
diff --git a/src/platform/qt/test/suite.h b/src/platform/qt/test/suite.h
index 6aae5f7e0..b5d40427f 100644
--- a/src/platform/qt/test/suite.h
+++ b/src/platform/qt/test/suite.h
@@ -12,12 +12,13 @@

 struct TestRegistrar {
        virtual int main(int argc, char** argv) = 0;
-};

-extern QList<TestRegistrar*> registeredTests;
+       static TestRegistrar* first;
+       TestRegistrar* next = nullptr;
+};

 #define M_REGISTER_QT_TEST(Class) \
        static struct Class ## Registrar : TestRegistrar { \
                QTEST_MAIN(Class) \
-               Class ## Registrar() { registeredTests << this; } \
+               Class ## Registrar() { next = first; first = this; } \
        } registrar_ ## Class;
```

Reply:
For documentation: You said you'd rather have one suite per executable, so this is moot and I'm just going to refactor this out.